apiserver/controllers/metrics.go

```diff
@@ -0,0 +1,93 @@
+package controllers
+
+import (
+	"log"
+
+	"garm/auth"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type GarmCollector struct {
+	instanceMetric *prometheus.Desc
+	apiController  *APIController
+}
+
+func NewGarmCollector(a *APIController) *GarmCollector {
+	return &GarmCollector{
+		apiController: a,
+		instanceMetric: prometheus.NewDesc(
+			"garm_runner_status",
+			"Status of the runner",
+			[]string{"name", "status", "runner_status", "pool", "pool_type", "hostname", "controller_id"}, nil,
+		)}
+}
+
+func (c *GarmCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.instanceMetric
+}
+
+func (c *GarmCollector) Collect(ch chan<- prometheus.Metric) {
+	c.CollectInstanceMetric(ch)
+}
+
+// CollectInstanceMetric collects the metrics for the runner instances
+// reflecting the statuses and the pool they belong to.
+func (c *GarmCollector) CollectInstanceMetric(ch chan<- prometheus.Metric) {
+
+	ctx := auth.GetAdminContext()
+
+	instances, err := c.apiController.r.ListAllInstances(ctx)
+	if err != nil {
+		log.Printf("cannot collect metrics, listing instances: %s", err)
+		return
+	}
+
+	pools, err := c.apiController.r.ListAllPools(ctx)
+	if err != nil {
+		log.Printf("listing pools: %s", err)
+		// continue anyway
+	}
+
+	type poolInfo struct {
+		Name string
+		Type string
+	}
+
+	poolNames := make(map[string]poolInfo)
+	for _, pool := range pools {
+		if pool.EnterpriseName != "" {
+			poolNames[pool.ID] = poolInfo{
+				Name: pool.EnterpriseName,
+				Type: "enterprise",
+			}
+		} else if pool.OrgName != "" {
+			poolNames[pool.ID] = poolInfo{
+				Name: pool.OrgName,
+				Type: "organization",
+			}
+		} else {
+			poolNames[pool.ID] = poolInfo{
+				Name: pool.RepoName,
+				Type: "repository",
+			}
+		}
+	}
+
+	hostname, controllerID := c.apiController.GetControllerInfo()
+
+	for _, instance := range instances {
+
+		ch <- prometheus.MustNewConstMetric(
```

Comment:
Does this have the potential to panic?

By convention, functions that contain ```Must``` usually panic if something is wrong. While this may be ok in ```init()``` functions, because you see that something is terribly wrong when you run the app, calling a function that can panic is not a good idea elsewhere. We would need to ```recover()``` somehow, or risk crashing garm.

Reply:
There is an equivalent without the ```Must``` bit. You can log the error and return/continue. But it must not panic.

Reply:
done